### PR TITLE
Allow hardware specific region-ACL extensions

### PR DESCRIPTION
**Pull request by meriac**
- Enables hardware specific region-ACL extensions 
- Move subregion configuration into HW-specific ACLs:
  - For K64F DMA Busmaster overrides are possible
  - For ARMv7-M Subregions can be defined- Enables access to box 0 memories by the Ethernet peripheral
- Background region privileges reduced to Core and Debugger

@AlessandroA

### CODE REVIEW

File: core/system/src/mpu/vmpu_freescale_k64_mem.c [hunk 1]
```diff
@@ -63,21 +63,27 @@ void vmpu_mem_switch(uint8_t src_box, uint8_t dst_box)
...
+     * - allow execution
+     * - give read access to ENET DMA bus master */
     vmpu_mpu_set_static_acl(1, FLASH_ORIGIN, (uint32_t) __uvisor_config.secure_end - FLASH_ORIGIN,
         UVISOR_TACL_SREAD |
         UVISOR_TACL_SEXECUTE |
         UVISOR_TACL_UREAD |
         UVISOR_TACL_UEXECUTE |
-        UVISOR_TACL_USER);
+        UVISOR_TACL_USER,
+        0x4UL<<18);
```
Comment (AlessandroA) [inline core/system/src/mpu/vmpu_freescale_k64_mem.c]: Could you add spaces in `0x4UL << 18)`? (also in the other occurrence below)

BTW, do we want to provide an interface/macro for selecting a certain bus master or to set the K64F-specific ACLs? The hardcoded numbers are not very user-friendly. If we don't want to do it now we can add and issue/story.

**Comment by meriac** [comment]
@AlessandroA I fixed the spacing problem. I would like to avoid though macros for now for the hw-specific ACLs, as they are currently only used in this two lines plus I don't want to make this "as thing".

File: core/system/src/mpu/vmpu_freescale_k64_mpu.c [hunk 2]
```diff
@@ -70,11 +70,17 @@ uint32_t vmpu_round_up_region(uint32_t addr, uint32_t size)
...
-static uint32_t vmpu_map_acl(UvisorBoxAcl acl)
+static uint32_t vmpu_map_acl(UvisorBoxAcl acl, uint32_t acl_hw_spec)
 {
     /* handle user permissions */
     uint32_t perm = (acl & UVISOR_TACL_USER) ? (acl & UVISOR_TACL_UACL) : 0;
 
+    /* Allow uVisor-internal code to extend permissions for certain
+     * ACLs by hardware-specific ACLs like DMA busmaster access.
+     * Ensure that core ACLs are specified trough the generic ACL
+     * method b masking it out. */
```
Comment (AlessandroA) [inline core/system/src/mpu/vmpu_freescale_k64_mpu.c]:
```diff
-    * method b masking it out. */
+    * method by masking it out. */
```

**Comment by AlessandroA** [comment]
LGTM 👍

**Comment by sherrellbc** [comment]
DMA confirmed to work (see https://github.com/ARMmbed/uvisor/issues/315) to/from box 0 using LwIP as a test case (the original motivation for this patch). Other problems related to managing mutually exclusive memory (mutex) at run-time have now manifested as either a side-effect of this patch or as something previously unknown and un-tested. See the above quoted git issue for more information.

**Pull request merged**